### ISSUE TITLE
Need QoS testing on Windows

### ISSUE THREAD

**Issue by PatrickLang** [kubernetes/kubernetes#73418]
https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/

We need a test case that:
1. Sets up a namespace with limits on CPU & memory
2. Creates multiple pods with differing QoS settings that will exceed the namespace limits
3. Remove a few pods at a time and make sure QoS behavior is preserved and that lower QoS pods do get scheduled when higher QoS ones are stopped

/sig windows
/milestone v1.14

**Comment by bclau** [comment]
So, this is a bit confusing, since QoS can refer to a few different things.

First of all, I was thinking about LimitRanges set per namespace, but the LimiRange types are: Pod, Container, PersistentVolumeClaim. The documentation says: ``If a Pod does not meet the constraints imposed by the LimitRange, it cannot be created in the namespace.``, which means that it doesn't fit the scenario described by your [2] point, as the LimitRange will be taken into account per Pod, not over all Pods in a Namespace.

So, the scenario described by [2] should be related to ResourceQuotas. In this regard, the test ``[sig-api-machinery] ResourceQuota should create a ResourceQuota and capture the life of a pod.`` already does what you need, plus a few extra things:

1. Creates a ResourceQuota and validates the ResourceQuota (should be empty).
2. Creates Pod that fits quota and validates the ResourceQuota updates accordingly.
3. Ensures that it cannot spawn a new pod which exceeds the quota (including with extended resources).
4. Ensures that it cannot update the original Pod's requested resources.
5. Ensures that the ResourceQuota did not change.
6. Deletes the pod and ensures that the ResourceQuota updates accordingly.

Other ResourceQuota related tests:


    [sig-api-machinery] ResourceQuota [Feature:PodPriority] should verify ResourceQuota's multiple priority class scope (quota set to pod count: 2) against 2 pods with same priority classes.
    [sig-api-machinery] ResourceQuota [Feature:PodPriority] should verify ResourceQuota's priority class scope (cpu, memory quota set) against a pod with same priority class.
    [sig-api-machinery] ResourceQuota [Feature:PodPriority] should verify ResourceQuota's priority class scope (quota set to pod count: 1) against 2 pods with different priority class.
    [sig-api-machinery] ResourceQuota [Feature:PodPriority] should verify ResourceQuota's priority class scope (quota set to pod count: 1) against 2 pods with same priority class.
    [sig-api-machinery] ResourceQuota [Feature:PodPriority] should verify ResourceQuota's priority class scope (quota set to pod count: 1) against a pod with different priority class (ScopeSelectorOpExists).
    [sig-api-machinery] ResourceQuota [Feature:PodPriority] should verify ResourceQuota's priority class scope (quota set to pod count: 1) against a pod with different priority class (ScopeSelectorOpNotIn).
    [sig-api-machinery] ResourceQuota [Feature:PodPriority] should verify ResourceQuota's priority class scope (quota set to pod count: 1) against a pod with same priority class.
    [sig-api-machinery] ResourceQuota [Feature:ScopeSelectors] should verify ResourceQuota with best effort scope using scope-selectors.
    [sig-api-machinery] ResourceQuota [Feature:ScopeSelectors] should verify ResourceQuota with terminating scopes through scope selectors.
    [sig-api-machinery] ResourceQuota should create a ResourceQuota and capture the life of a configMap.
    [sig-api-machinery] ResourceQuota should create a ResourceQuota and capture the life of a persistent volume claim with a storage class. [sig-storage]
    [sig-api-machinery] ResourceQuota should create a ResourceQuota and capture the life of a persistent volume claim. [sig-storage]
    [sig-api-machinery] ResourceQuota should create a ResourceQuota and capture the life of a pod.
    [sig-api-machinery] ResourceQuota should create a ResourceQuota and capture the life of a replica set.
    [sig-api-machinery] ResourceQuota should create a ResourceQuota and capture the life of a replication controller.
    [sig-api-machinery] ResourceQuota should create a ResourceQuota and capture the life of a secret.
    [sig-api-machinery] ResourceQuota should create a ResourceQuota and capture the life of a service.
    [sig-api-machinery] ResourceQuota should create a ResourceQuota and ensure its status is promptly calculated.
    [sig-api-machinery] ResourceQuota should verify ResourceQuota with best effort scope.
    [sig-api-machinery] ResourceQuota should verify ResourceQuota with terminating scopes.

Also, for behaviour [3], normally, it is not possible. If a new pod will exceed the set ResourceQuota, the Kubernetes API will error out instead, and no other pod is created:

    kubectl create -f pods/busybox_windows_pod_res.yaml -n test-rq
    Error from server (Forbidden): error when creating "pods/busybox_windows_pod_res.yaml": pods "win-busybox2" is forbidden: exceeded quota: mem-cpu-demo, requested: requests.cpu=1,requests.memory=1Gi, used: requests.cpu=1,requests.memory=1Gi, limited: requests.cpu=1,requests.memory=1Gi

One way [3] could work, would be when ``PodPriority`` and ``PodPremption`` features are being used. And for that, there are a few tests:

    [sig-scheduling] SchedulerPreemption [Serial] validates basic preemption works
    [sig-scheduling] SchedulerPreemption [Serial] validates lower priority pod preemption by critical pod
    [sig-scheduling] SchedulerPreemption [Serial] validates pod anti-affinity works in preemption

**Comment by adelina-t** [comment]
@adelina-t

**Comment by brendandburns** [comment]
There are some tests here:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/pods_container_manager_test.go
https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_manager_test.go

The CGroup tests are very Linux specific. I don't think they actually apply to Windows because on Windows there's no such thing as the kernel OOM killer, and thus Memory is actually compressible on Windows (see definitions here: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/resource-qos.md#requests-and-limits)

I believe that all of the ResourceQuota tests should work unchanged because they are enforced in the API Server not on the node. (we may need to adjust the image and the command for those tests)

So basically if we ensure that we're reporting resources on the node correctly, I think we have covered this ask.

**Comment by immuzz** [comment]
FYI I'm working on this and running ResourceQuota tests against Windows. I'll post updates. @PatrickLang

**Comment by timmycarr** [comment]
Hello,

I'm a bug triage shadow and I'd like to remind that code freeze is coming in 3 days (Friday, 8th March). Is it still planned for this issue to be fixed for 1.14?

**Comment by PatrickLang** [comment]
The right tests were added in https://github.com/kubernetes/test-infra/pull/11301.

/close
